### ISSUE TITLE
Slower uniq in materialized view compared to normal query on base table

### ISSUE THREAD

**Issue by a-dot** [ClickHouse/ClickHouse#29823]
Title says it all..

Reproducable by following this:
(using ClickHouse server version 21.9.1 revision 54449)

```
CREATE TABLE IF NOT EXISTS nf
(
    `clientIP` String,
    `serverIP` String
)
ENGINE = MergeTree
ORDER BY (serverIP, clientIP)
```
```
CREATE MATERIALIZED VIEW nf_uniq_mv
ENGINE = AggregatingMergeTree
ORDER BY (serverIP)
AS
SELECT
    serverIP,
    uniqState(clientIP) as uniq_clientIP
FROM nf
GROUP BY (serverIP)
```

```
INSERT INTO nf (clientIP, serverIP)
SELECT
    IPv4NumToString(toUInt32(toUInt32(16777216 + rand(2) % 16777215))),
    IPv4NumToString(toUInt32(toUInt32(16908288 + rand(2) % 65535)))
FROM numbers_mt(100000000)
```

Query the materialized view:
`select serverIP, uniqMerge(uniq_clientIP) as uniq_cips from nf_uniq_mv GROUP BY (serverIP);`
Result: 65535 rows in set. Elapsed: 33.561 sec. Processed 131.07 thousand rows, 20.30 MB (3.91 thousand rows/s., 604.81 KB/s.)

Query the base table:
`select serverIP, uniq(clientIP) as uniq_cips from nf GROUP BY serverIP;`
Result: 65535 rows in set. Elapsed: 5.259 sec. Processed 100.00 million rows, 3.99 GB (19.01 million rows/s., 757.75 MB/s.)

Querying the base table results in processing 99,868,930 more rows than querying the materialized view. Yet, it takes 28.302 LONGER to get the same results back.

I would expect the materialized view to be faster than querying the base table.

**Comment by a-dot** [comment]
I'm getting a very different result than you are. I tried what you suggested, here are my results:

```
axiom :) set max_threads=1;

SET max_threads = 1
```

```
axiom :) select serverIP, uniqMerge(uniq_clientIP) as uniq_cips from nf_uniq_mv GROUP BY (serverIP) format Null;

Ok.

0 rows in set. Elapsed: 13.432 sec. Processed 131.07 thousand rows, 20.30 MB (9.76 thousand rows/s., 1.51 MB/s.)
```

```
aphorism :) select serverIP, uniq(clientIP) as uniq_cips from nf GROUP BY serverIP format Null;

Ok.

0 rows in set. Elapsed: 8.914 sec. Processed 100.00 million rows, 3.99 GB (11.22 million rows/s., 447.07 MB/s.)
```

My materialized view query is still significantly slower.

**Comment by PyCht** [comment]
@den-crane I see same results on my testing using clickhouse 22.8
I'm comparing MV with uniqState to uniq on raw data, no OS caching: min_bytes_to_use_direct_io=1

uniqMerge: 20 rows in set. Elapsed: 3.357 sec. Processed 102.47 million rows, 18.55 GB (30.52 million rows/s., 5.52 GB/s.)
uniq: 20 rows in set. Elapsed: 3.276 sec. Processed 170.15 million rows, 17.12 GB (51.94 million rows/s., 5.23 GB/s.)

It looks like *Merge functions take much time to complete, is this what you expect?
Is there a ratio where it would be more preferment to use the Merge notation instead of just querying the raw data?

**Comment by SaltTan** [comment]
Have you tried `select serverIP, uniqMerge(uniq_clientIP) as uniq_cips from nf_uniq_mv GROUP BY serverIP`?

**Comment by PyCht** [comment]
> 

Yes, this is what I was thinking, i saw significant improvement when MV has only 10% of the rows.
State Aggregation Functions stores the "state" in disk so the number pf processed must be much lower to see significant improvement using such functions.
